### ISSUE TITLE
Bug in losing 'next' field of header stack during sub-parser invocation

### ISSUE THREAD

**Issue by jafingerhut** [p4lang/p4c#1128]
I don't have a test case for this right now.  This is more of a case of seeing a bug in one place, where probably soon it will be fixed, but wondering whether a similar bug may also exist in another place.

Background: I am writing a P4_16 test program to test the new P4_16 header stack support made by this behavioral-model PR: https://github.com/p4lang/behavioral-model/pull/515

I started by writing a program and testing with those new changes, and then I wanted to verify that some of those tests, ones that should fail with the old/current behavior of 'push', actually do fail.

My P4_16 program has an ingress control block that passes all headers to a sub-control, modifying the headers there, then returning to the ingress control.  The sub-control does P4_16 push_front and pop_front() method calls on a header stack, passed as a field of a 'struct headers' type that is commonly seen in P4_16 programs.

The BMv2 JSON file compiled from this source code has 2 header stacks -- one for the 'global' one in the ingress control block, and another temporary one for the sub-control.  The JSON file has actions that when calling the sub-control, copies the elements of the size-5 header stack via 5 'assign_header' primitive actions from the global header stack to the temporary one, then copies them back out when returning.

That is correct for getting most of the state of the header stack from the ingress control to the sub-control being called, but it _does not_ cause the 'next' state of the header stack to be copied in and out on these calls to the sub-control.

Now maybe this is soon to be a historical footnote for control blocks, because with the new behavior of push and pop primitives, they will no longer depend upon the value of 'next' in a header stack.

However, if the compiler can generate similar BMv2 JSON code for a parser calling a sub-parser, then it is possible it can generate code that does not behave as the source code specifies, because of losing the 'next' state.

**Comment by jafingerhut** [comment]
This is likely only an issue for P4_16 programs, since I am guessing that for P4_14 programs having all headers and header stacks global leaves no reason for the compiler to add copying of one header stack to another.

I don't think P4_14 allows copying one header stack variable to another, but P4_16 does, and if p4c implements such copying in a parser, that might be another situation where incorrect behavior could occur due to not copying the 'next' state from one header stack to another.  I haven't tested that.

**Comment by jafingerhut** [comment]
Note: Perhaps a reasonable way to fix this for existing bmv2 header stack support, and for the future, is to have a primitive that assigns one header stack to another, which copies all headers in the stack, which is already being copied, plus the 'next' state that is currently not being copied.  That seems like it should solve the existing problems I've seen in the control block calls, plus also if that technique is used for parser to sub-parser calls, and make it easier to preserve any other state that might be 'hidden' inside of a header stack.

The same approach could help for header union copying, and stack of header union copying, if/when those things are copied around.  Perhaps something like that is already done -- I haven't dug into the details there yet.

**Comment by antoninbas** [comment]
There is already a "copy" primitive for header unions (https://github.com/p4lang/behavioral-model/blob/master/include/bm/bm_sim/core/primitives.h#L46).
I can definitely try to add one for header stacks and header union stacks. However, these copies are fairly expensive and I would recommend avoiding them when possible.

**Comment by jafingerhut** [comment]
I do not know the p4c internals well enough, but I believe this copying is only generated by the compiler during calls to sub-controls, and even then it depends on the parameters to the sub-control whether the compiler generates explicit copy-in and copy-out code in the BMv2 JSON file, or simply shares the same variables that the caller is using.

There are cases where p4c can generate very inefficient code, compared to what a person would do, when calling sub-controls or global actions: https://github.com/p4lang/p4c/issues/599

At least having correct-but-inefficient execution in bmv2 would probably be preferable to subtly-incorrect-but-fast execution.

**Comment by jafingerhut** [comment]
I do have an example program available to look at.  It is the one added by this PR: https://github.com/p4lang/p4c/pull/1129

The comments in the STF file show which tests behave correctly, vs. which exhibit bugs in the bmv2 JSON file generated by the latest version of p4c-bm2-ss as of 2017-Jan-05.

**Comment by mbudiu-vmw** [comment]
I regret the choice made for the design of header stacks of having the "next" field be updated by calls to extract. This is causing lots of problems. We should have kept the previous design where you can infer 'next' by looking at the valid elements in the stack. Then you don't need a hidden field in the array to implement arrays.

The 'next' field prevents us from doing source-to-source translation to implement several array-related operations, including array assignment, since we cannot express any operations on the 'next' field in P4 source code. 

Unfortunately we can also not synthesize a separate scalar to implement the behavior of 'next' explicitly, because some targets will allow array.next but not array[next], where next is not a compile-time constant. 
Does bmv2 allow indexing into an array with a non-constant value?

**Comment by jafingerhut** [comment]
@mbudiu-vmw Would this issue be straightforward to fix if bmv2 had a primitive for copying one header stack to another, which included copying the 'next' state?

At least for the cases of incorrect generated code I have seen so far, using such a primitive should take about the same amount of time as the code that is being generated now, which is to do N assign_header calls to copy all-except-the-next-field from one header stack to another.

**Comment by mbudiu-vmw** [comment]
This would work for bmv2. 
But in general the compiler will be unable to lower array copy to a set of more elementary operations.

**Comment by jafingerhut** [comment]
Got it.  Thanks for saying the same thing in a different way so that I realized my question was ignoring other P4 backends that p4c should be able to support.

**Comment by jafingerhut** [comment]
I have created a test case that gives incorrect output because of this issue, one that uses a header stack and extract operations on the header stack only, no push/pop operations.  https://github.com/p4lang/p4c/pull/1132

**Comment by jafingerhut** [comment]
@mbudiu-vmw 

Section 8.15 of the P4_16 language spec says that one header stack can be assigned to another, and gives more details about what such an assignment means:

"assignment from a header stack hs into another stack requires the stacks to have the same types and sizes. All components of hs are copied, including its elements and their validity bits, as well as nextIndex. "

In this comment above: https://github.com/p4lang/p4c/issues/1128#issuecomment-356034889

you say: "The 'next' field prevents us from doing source-to-source translation to implement several array-related operations, including array assignment, since we cannot express any operations on the 'next' field in P4 source code."

It seems that array assignment can be expressed in P4_16 source code, according to the spec, and it does copy that value of next (called nextIndex in the spec).

What am I missing?

**Comment by jafingerhut** [comment]
Quick hack experiment: Commenting out the branch of the if statement beginning with the line "else if (ltype->is<IR::Type_Stack>())" in the postorder method of file p4c/midend/copyStructures.cpp prevents assignments between header stacks, even ones generated by an earlier step of the compiler itself, from being expanded into a sequence of N assignments between the N header elements of the stack.  The BMv2 backend code produces a "set" primitive between the two header stacks, which bmv2 code does not implement and gives an error for.

If bmv2 either implemented that case for the "set" primitive, that sounds like one way to fix this bug.

**Comment by mbudiu-vmw** [comment]
@antoninbas : I have written a very simple program using array assignment, but it does not seem to produce the expected result: the second array that is created by copying is full of zeros when emitted. I will attach 3 files: the source, the stf test case and the json. I have some code ready which fixes the code generation to use the array assignment operation, but it does not pass this test. The current version of the compiler will not generate this json. The core of the program is this:
```
 state start {
        p.extract(h.h1.next);
        p.extract(h.h1.next);
        h.h2 = h.h1;
        transition accept;
    }
```


[array-copy.zip](https://github.com/p4lang/p4c/files/1644971/array-copy.zip)

**Comment by antoninbas** [comment]
@mbudiu-vmw Thanks for catching this. It was a one line fix. It should be fixed in master once this is merged: https://github.com/p4lang/behavioral-model/pull/534

**Comment by jafingerhut** [comment]
I haven't read through every comment in this thread, but looks like this issue is fixed now that PR #1144 is merged.  Feel free to close this one.  If there is not-yet-done work mentioned in comments above that needs to be tracked, probably best to open separate issue for it.
